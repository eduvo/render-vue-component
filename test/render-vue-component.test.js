/**
 * @jest-environment jsdom
 */
import flushPromises from "flush-promises"
import Vue from "vue"

describe('RenderVueComponent', () => {
  let RenderVueComponent

  beforeEach(async () => {
    module = await import("../src/index")
    RenderVueComponent = module.default
    jest.resetModules()
  })

  describe('private settings Vue', () => {
    it('should set Vue to private settings.vue by setVue(vue)', () => {
      RenderVueComponent.setVue("test")
      expect(RenderVueComponent.getVue()).toBe("test")
    });

    it('should return settings.vue by getVue()', () => {
      expect(RenderVueComponent.getVue()).toEqual(Vue)
    });

    it('should accept function as parameter and can manipulate Vue inside the function', () => {
      RenderVueComponent.setVue({ a: 1 })
      RenderVueComponent.getVue((vue) => { vue.a = 2 })
      expect(RenderVueComponent.getVue().a).toBe(2)
    });
  });

  describe('register()', () => {
    it('should register the component by settings.Vue', () => {
      RenderVueComponent.setVue(Vue)
      RenderVueComponent.register({ 
        'BaseTag': 'BaseTagComponent',
        'BaseTag2': 'BaseTagComponent2',
      })
      expect(Vue.component).toHaveBeenCalledWith('BaseTag', 'BaseTagComponent')
      expect(Vue.component).toHaveBeenCalledWith('BaseTag2', 'BaseTagComponent2')

      Vue.component.mockReset()
    });
  });

  describe('renderComponentsOnRails()', () => {
    beforeEach(() => {
      Vue.component('BaseTag', {
        props: ['text'],
        template: '<div class="base-tag">{{ text }}</div>'
      })
      Vue.component('BaseButton', {
        props: ['text'],
        template: '<button class="base-button">{{ text }}</button>'
      })
      RenderVueComponent.setVue(Vue)

      document.body.innerHTML =
        '<div ' +
          'class="vue-component"' +
          'data-name="BaseTag" ' +
          'data-rails-data=\'{"text": "Test1"}\' ' +
        '></div>' +
        '<div ' +
          'class="vue-component" ' +
          'data-name="BaseButton" ' +
          'data-rails-data=\'{"text": "Test2"}\' ' +
        '></div>'
    })

    it('should mount Vue component specified by element data-name to .vue-component element', async () => {
      RenderVueComponent.renderComponentsOnRails()
      await flushPromises()
      expect(document.body).toMatchSnapshot()
    });
  });
});