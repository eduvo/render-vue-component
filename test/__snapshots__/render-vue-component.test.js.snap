// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderVueComponent renderComponentsOnRails() should mount Vue component specified by element data-name to .vue-component element 1`] = `
<body>
  <!--function (a, b, c, d) { return createElement(vm, a, b, c, d, true); }-->
  <!--function (a, b, c, d) { return createElement(vm, a, b, c, d, true); }-->
</body>
`;
